cmd/gpu_plugin/rm/gpu_plugin_resource_manager.go

```diff
@@ -0,0 +1,417 @@
+		for _, id := range crqt.DevicesIDs {
+			if id == monitorID {
+				return false // intentionally not printing anything
+			}
+		}
+	}
+
+	return true
+}
+
+// findAllocationPodCandidate tries to find the best allocation candidate pod, which must be:
+//  -pending for this node
+//  -using GPU resources in its spec
+//  -is found via grpc service with unallocated GPU devices
+// returns:
+//  -the candidate pod struct pointer and no error, or
+//  -errRetry if unsuccessful, but there is perhaps hope of trying again with better luck
+//  -errZeroPending if no pending pods exist anymore (which is fine)
+//  -any grpc communication errors
+func (rm *ResourceManager) findAllocationPodCandidate() (*podCandidate, error) {
+	// get map of pending pods for this node
+	pendingPods, err := rm.getNodePendingGPUPods()
+	if err != nil {
+		return nil, errors.Wrap(err, "getting pending pods failed")
+	}
+
+	var candidate podCandidate
+	candidates, err := findAllocationPodCandidates(pendingPods)
+	if err != nil {
+		return nil, errors.Wrap(err, "finding allocation pod candidates failed")
+	}
+
+	numCandidates := len(candidates)
+	switch numCandidates {
+	case 0:
+		// fine, this typically happens when deployment is deleted before PODs start
+		klog.V(4).Info("zero pending pods")
+		return nil, errZeroPending
+	case 1:
+		// perfect, only one option
+		klog.V(4).Info("only one pending pod")
+		candidate = candidates[0]
+		if _, ok := candidate.pod.Annotations[gasCardAnnotation]; !ok {
+			klog.Warningf("Pending POD annotations from scheduler not yet visible for pod %v", candidate.pod.Name)
+			return nil, errRetry
+		}
+	default: // > 1 candidates, not good, need to pick the best
+		// look for scheduler timestamps and sort by them
+		klog.V(4).Infof("%v pods pending, picking oldest", numCandidates)
+		timestampedCandidates := []podCandidate{}
+		for _, candidate := range candidates {
+			if _, ok := pendingPods[candidate.name].Annotations[gasTSAnnotation]; ok {
+				timestampedCandidates = append(timestampedCandidates, candidate)
+			}
+		}
+		sort.Slice(timestampedCandidates,
+			func(i, j int) bool {
+				return pendingPods[timestampedCandidates[i].name].Annotations[gasTSAnnotation] <
+					pendingPods[timestampedCandidates[j].name].Annotations[gasTSAnnotation]
+			})
+		if len(timestampedCandidates) == 0 {
+			klog.Warning("Pending POD annotations from scheduler not yet visible")
+			return nil, errRetry
+		}
+		candidate = timestampedCandidates[0]
+	}
+
+	candidate.pod = pendingPods[candidate.name]
+	return &candidate, nil
+}
+
+// +kubebuilder:rbac:groups="",resources=pods,verbs=list
+
+// getNodePendingGPUPods returns a map of pod names -> pods that are pending and use the gpu.
+func (rm *ResourceManager) getNodePendingGPUPods() (map[string]*v1.Pod, error) {
+	selector, err := fields.ParseSelector("spec.nodeName=" + rm.nodeName +
+		",status.phase=" + string(v1.PodPending))
+
+	if err != nil {
+		klog.Error(err.Error())
```

Comment:
Currently the error is logged twice: here and in the line 127. Better drop this line and log errors at the place where they are handled.